platform/jewel/markdown/core/src/main/kotlin/org/jetbrains/jewel/markdown/rendering/DefaultMarkdownBlockRenderer.kt

```diff
@@ -439,6 +472,46 @@ public open class DefaultMarkdownBlockRenderer(
             inlineRenderer.renderAsAnnotatedString(block.inlineContent, styling, enabled, onUrlClick)
         }
 
+    @Composable
+    private fun renderedImages(blockInlineContent: List<InlineMarkdown>): Map<String, InlineTextContent> =
+        getImages(blockInlineContent).associate { image -> image.source to imageContent(image) }
+
+    @Composable
+    private fun imageContent(image: InlineMarkdown.Image): InlineTextContent {
+        val knownSize = remember(image.source) { mutableStateOf<ImageSize?>(null) }
+        return InlineTextContent(
+            with(LocalDensity.current) {
+                // `toSp` ensures that the placeholder size matches the original image size in
+                // pixels.
+                // This approach doesn't allow images from appearing larger with different screen
+                // scaling,
+                // but simply maintains behavior consistent with standalone AsyncImage rendering.
+                Placeholder(
+                    width = knownSize.value?.width?.dp?.toSp() ?: 0.sp,
+                    height = knownSize.value?.height?.dp?.toSp() ?: 1.sp,
+                    placeholderVerticalAlign = PlaceholderVerticalAlign.Bottom,
+                )
+            }
+        ) {
+            AsyncImage(
+                model =
+                    ImageRequest.Builder(LocalPlatformContext.current)
+                        .data(image.source)
+                        // make sure image doesn't get downscaled to the placeholder size
+                        .size(Size.ORIGINAL)
+                        .build(),
+                contentDescription = image.title,
+                onSuccess = { state ->
+                    if (knownSize.value == null) {
```

Comment:
Is the check needed? I'd imagine whatever we get in onSuccess we only get once anyway...